scalanet/test/src/io/iohk/scalanet/peergroup/kademlia/BitVectorCodecSpec.scala

```diff
@@ -0,0 +1,78 @@
+package io.iohk.scalanet.peergroup.kademlia
+
+import java.nio.ByteBuffer
+
+import io.iohk.decco.Codec
+import org.scalacheck.Arbitrary
+import org.scalacheck.Arbitrary.arbitrary
+import org.scalactic.Equivalence
+import org.scalatest.FlatSpec
+import scodec.bits.BitVector
+
+import scala.util.Random
+
+class BitVectorCodecSpec extends FlatSpec {
+
+  import BitVectorCodec._
+  import BitVectorCodecSpec._
+  import io.iohk.decco.auto._
+
+  implicit val arbitraryBitVector = Arbitrary(Generators.genBitVector())
+
+  "BitVectorCodec" should "work" in testCodec[BitVector]
+}
+
+object BitVectorCodecSpec {
+  import io.iohk.decco.BufferInstantiator.global.HeapByteBuffer
+  import org.scalatest.EitherValues._
+  import org.scalatest.Matchers._
+  import org.scalatest.prop.GeneratorDrivenPropertyChecks._
+
+  private def encodeDecodeTest[T](implicit codec: Codec[T], a: Arbitrary[T], eq: Equivalence[T]): Unit = {
+    forAll(arbitrary[T]) { t =>
+      val e = codec.encode(t)
+      codec.decode(e).right.value should equal(t)
+    }
+  }
+
+  private def variableLengthTest[T](implicit codec: Codec[T], a: Arbitrary[T], eq: Equivalence[T]): Unit = {
+    forAll(arbitrary[T]) { t =>
+      // create a buffer with one half full of real data
+      // and the second half full of rubbish.
+      // codecs should not be fooled by this.
+      val b = codec.encode(t).toArray
+      val newB = ByteBuffer
+        .allocate(b.length * 2)
+        .put(b)
+        .put(randomBytes(b.length))
+      (newB: java.nio.Buffer).position(0)
+
+      codec.decode(newB).right.value shouldBe t
```

Comment:
Interesting, but this kinda looks like something to add to decco and not here.
At the same time, is this actually the expected behaviour? Could you expand on the reason?

Reply:
The test is checking that, when decoding, a codec does not read beyond the end of the bytes that were encoded.